src/wp-includes/html-api/class-wp-xml-decoder.php

```diff
@@ -0,0 +1,94 @@
+<?php
+
+/**
+ * XML API: WP_XML_Decoder class
+ *
+ * Decodes spans of raw text found inside XML content.
+ *
+ * @package WordPress
+ * @subpackage HTML-API
+ * @since WP_VERSION
+ */
+class WP_XML_Decoder {
+
+	/**
+	 * Decodes a span of XML text.
+	 *
+	 * Example:
+	 *
+	 *     '©' = WP_XML_Decoder::decode( 'data', '&copy;' );
+	 *
+	 * @since WP_VERSION
+	 *
+	 * @access private
+	 *
+	 * @param string $text    Text document containing span of text to decode.
+	 * @return string Decoded UTF-8 string.
+	 */
+	public static function decode( $text ) {
+		$decoded = '';
+		$end     = strlen( $text );
+		$at      = 0;
+		$was_at  = 0;
+
+		while ( $at < $end ) {
+			$next_character_reference_at = strpos( $text, '&', $at );
+			if ( false === $next_character_reference_at || $next_character_reference_at >= $end ) {
+				break;
+			}
+
+			// Use the `attribute` context to disable ambiguous follower detection.
+			$character_reference = WP_HTML_Decoder::read_character_reference( 'attribute', $text, $next_character_reference_at, $token_length );
```

Comment:
I think we’ll want to only decide the five mandated character references plus numeric character references, meaning using the HTML decoder is going to be wrong and slow here.

I don’t know the state of spec compliance, but I believe that the ironically named `html_entity_decode()` is better here than the HTML device would be.

best is probably the numeric decoder plus trapping those five entities defined in XML

Reply:
Done in 1d4026b